src/rulesets/Base/mapreduce.jl

```diff
@@ -67,30 +67,55 @@ function rrule(
 end
 
 function rrule(
-    config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f, xs::AbstractArray; dims=:
-)
-    fx_and_pullbacks = map(x->rrule_via_ad(config, f, x), xs)
-    y = sum(first, fx_and_pullbacks; dims=dims)
+    config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f::F, xs::AbstractArray; dims=:
+) where {F}
+    project = ProjectTo(xs)
 
-    pullbacks = last.(fx_and_pullbacks)
+    if _uses_input_only(f, eltype(xs))
+        # Then we can compute the forward pass as usual, save nothing but `xs`:
+        y = sum(f, xs; dims=dims)
+        function sum_pullback_easy(dy)
+            dxs = unthunk(dy) .* conj.(only.(only.(derivatives_given_output.(nothing, f, xs))))
+            return (NoTangent(), NoTangent(), project(dxs))
+        end
+        return y, sum_pullback_easy
+    end
 
-    project = ProjectTo(xs)
+    # In the general case, we need to save all the pullbacks:
+    fx_and_pullbacks = map(x -> rrule_via_ad(config, f, x), xs)
+    y = sum(first, fx_and_pullbacks; dims=dims)
+
+    function sum_pullback_f(dy)
+        # For arrays of arrays, we ought to protect the element against broadcasting:
+        dys = dims isa Colon ? Ref(unthunk(dy)) : unthunk(dy)
+        if Base.issingletontype(F)
+            # Then at least `f` has no gradient. Note that broadcasting here
+            # gets the shape right with or without `dims` keyword.
+            dxs = broadcast(fx_and_pullbacks, dys) do (_, back), dy1
+                unthunk(last(back(dy1)))
```

Comment:
I don't think the `unthunk` is required anymore, as didn't we fix `project`?

```suggestion
                last(back(dy1))
```

Reply:
I think my test case here was `sum(sum, ...)` where you get an InplaceableThunk from `back` here.

Reply:
And `ProjectTo` does not like those:
```
julia> proj = ProjectTo([1,2,3]);

julia> ith = rrule(sum, [1,2,3])[2](1)[2]; ith isa InplaceableThunk
true

julia> proj(ith)
ERROR: MethodError: no method matching (::ProjectTo{AbstractArray, NamedTuple{(:element, :axes), Tuple{ProjectTo{Float64, NamedTuple{(), Tuple{}}}, Tuple{Base.OneTo{Int64}}}}})(::InplaceableThunk{Thunk{ChainRules.var"#1409#1412"{Int64, Colon, Vector{Int64}, ProjectTo{AbstractArray, NamedTuple{(:element, :axes), Tuple{ProjectTo{Float64, NamedTuple{(), Tuple{}}}, Tuple{Base.OneTo{Int64}}}}}}}, ChainRules.var"#1408#1411"{Int64, Colon}})
```
Although I think at some point I wanted to make it un-wrap, insert itself into the `@thunk` part, and return that.

In which case you'd get an array of thunks back, not an array of arrays. Not sure what we think about that.

Reply:
The first test which fails without this `unthunk` is:
```julia
julia> test_rrule(sum, sum, [[1,2], [3,4], [5,6]]; check_inferred=false)
test_rrule: sum on typeof(sum),Vector{Vector{Int64}}: Test Failed at /Users/me/.julia/packages/ChainRulesTestUtils/XI7i2/src/testers.jl:307
  Expression: ad_cotangent isa NoTangent
   Evaluated: Thunk{ComposedFunction{ProjectTo{AbstractArray, NamedTuple{(:element, :axes), ...

julia> CFG = ChainRulesTestUtils.ADviaRuleConfig();

julia> rrule(CFG, sum, sum, [[1,2], [3,4], [5,6]])[2](1.0)
(NoTangent(), NoTangent(), Thunk{ComposedFunction{ProjectTo{AbstractArray, NamedTuple{(:element, :axes)   
```
This works on the tagged version, something unthunks:
```
julia> test_rrule(sum, sum, [[1,2], [3,4], [5,6]]; check_inferred=false)
Test Summary:                                        | Pass  Total  Time
test_rrule: sum on typeof(sum),Vector{Vector{Int64}} |    7      7  0.8s
Test.DefaultTestSet("test_rrule: sum on typeof(sum),Vector{Vector{Int64}}", Any[], 7, false, false, true, 1.645230607920931e9, 1.645230608709222e9)

julia> CFG = ChainRulesTestUtils.ADviaRuleConfig();

julia> rrule(CFG, sum, sum, [[1,2], [3,4], [5,6]])[2](1.0)
(NoTangent(), NoTangent(), [[1.0, 1.0], [1.0, 1.0], [1.0, 1.0]])

(@v1.8) pkg> st ChainRules
Status `~/.julia/environments/v1.8/Project.toml`
⌃ [082447d4] ChainRules v1.26.0
```